### ISSUE TITLE
Allow Log Following/Consuming

### ISSUE THREAD

**Issue by ClaytonNorthey92** [testcontainers/testcontainers-go#133]
**Describe the bug**
Not a bug, a feature request.
When running containers via testcontainers-go, in order to view the `stdout` from a container you have to call `.Logs(...)` on that container.  I think it would be useful allow an option to forward the `stdout` output from the containers to some sort of log consumer.  For example, you could send this to one that would log `stdout` of the container to `stdout` for the host machine.  This would help with debugging.

The Java implementation uses [follow logs](https://www.testcontainers.org/features/container_logs/#streaming-logs)

**To Reproduce**
Not bug bug, no reproduction steps.

**Expected behavior**
Allow an option to be passed into the `ContainerRequest` to forward `stdout` to a follower, maybe...

```go
type LogConsumer interface {
    Accept(string logMessage) error
}
```

```go
type ContainerRequest struct {
	...
        LogFollowers []LogConsumer
}
```

```go
// an example consumer
type StdoutLogConsumer struct {}

func (s *StdoutLogConsumer) Accept(string logMessage) error {
    fmt.Println(logMessage)
}
```
**Additional context**
This was inspired by [this comment](https://github.com/testcontainers/testcontainers-go/issues/118#issuecomment-567487947).

**Comment by gianarb** [comment]
There is a followoutput method in the testcontainers library:

https://www.testcontainers.org/features/container_logs/

I think it will be good to have something like that. I am not sure if there are interfaces in the std library we can leverage, I am thinking about the `io.Reader` one for example https://golang.org/pkg/io/#Reader to stay a bit more generic and closer to what the community expects

**Comment by ClaytonNorthey92** [comment]
The [Java implementation](https://www.testcontainers.org/features/container_logs/) allows you to consume `stdout`, `stdin`, or both.

I think the best way to achieve this would be to use the Docker API, I don't think we need another library to handle this.

The Docker API itself allows [a "follow" option](https://docs.docker.com/engine/api/v1.30/#operation/ContainerLogs)

This maps to a ["follow" parameter in the Go client when calling ContainerLogs()](https://godoc.org/github.com/docker/docker/api/types#ContainerLogsOptions)

I think the correct thing to do here is to allow a user to call a `FollowOutput` method just like the Java implementation, pass in their consumer and another parameter that would (again, like the Java implementation) specify `stdin`, `stdout`, or both.  This would just call the `ContainerLogs` method with the `Follow` parameter set to `true`.  This would return an `io.ReadCloser` which we could use to pass logs to the consumer.

I am not sure of the exact details yet, I would have to dig into it more, but I think that is where I would want to start.

**Comment by ClaytonNorthey92** [comment]
@gianarb 

I was investigating this a bit, and if we pass `Follow: true` when we call `ContainerLogs`, the `io.ReadCloser` that is returned will behave like so:

when calling `.Read` for the first time, you will get a header in the form of `[8]byte`, this will ultimately give you the size of the line in the log
when you call `.Read` again, you should use a `[]byte` with size equal to that from the header returned in the previous step

you continue to perform those two steps for each line in the log, that is how you can "follow" output
[this post answer](https://stackoverflow.com/a/46428872/4227700) explains it well

so I am wondering how we want to approach this, the Java implementation uses a [Consumer](https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html) which it will pass an [output frame](https://github.com/testcontainers/testcontainers-java/blob/f113979ad5207464490cc4f19f488bb27795802c/core/src/main/java/org/testcontainers/containers/output/OutputFrame.java) to the Consumer, which looks just to be an object with a content (bytes) and type (stdin, stdout, stderr).

so should we try to do the same abstraction?  should each of the consumers we create simply accept a struct similar to
```go
type OutputFrame struct {
    Type string
    Content []byte
}
```

and each container would have some sort of log follower concurrent process (a goroutine) that would always read up to the latest line in the log file and pass it to all of the consumers as `OutputFrame`s?  so, our `FollowOutput` would simply add a consumer to a slice of consumers, who would each get each log line

sorry if that is a lot, I can answer questions if it doesn't make sense

**Comment by ClaytonNorthey92** [comment]
@gianarb I will get an example pull request in to show you what I am thinking

**Comment by ClaytonNorthey92** [comment]
hey @gianarb I opened up a draft pull request #135  to proof out how I think this could work, there are a few things I want to add before I no longer call it a "draft".  I have left comments in the pull request for those.  could you take a look?

**Comment by gianarb** [comment]
Closed #135
